flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java

```diff
@@ -82,11 +113,58 @@ public JobGraph retrieveJobGraph(Configuration configuration) throws FlinkExcept
 	}
 
 	private PackagedProgram createPackagedProgram() throws FlinkException {
+		final String entryClass = getJobClassNameOrScanClassPath();
 		try {
-			final Class<?> mainClass = getClass().getClassLoader().loadClass(jobClassName);
+			final Class<?> mainClass = getClass().getClassLoader().loadClass(entryClass);
 			return new PackagedProgram(mainClass, programArguments);
 		} catch (ClassNotFoundException | ProgramInvocationException e) {
 			throw new FlinkException("Could not load the provided entrypoint class.", e);
 		}
 	}
+
+	private String getJobClassNameOrScanClassPath() throws FlinkException {
+		if (jobClassName != null) {
+			return jobClassName;
+		}
+
+		try {
+			return scanClassPathForJobJar();
+		} catch (IOException | NoSuchElementException | IllegalArgumentException e) {
+			throw new FlinkException("Failed to find job JAR on class path. Please provide the job class name explicitly.", e);
+		}
+	}
+
+	private String scanClassPathForJobJar() throws IOException {
+		LOG.info("Scanning class path for job JAR");
+		JarFileWithEntryClass jobJar = JarManifestParser.findOnlyEntryClass(jarsOnClassPath.get());
+
+		LOG.info("Using {} as job jar", jobJar);
+		return jobJar.getEntryClass();
+	}
+
+	@VisibleForTesting
+	enum JarsOnClassPath implements Supplier<Iterable<File>> {
+		INSTANCE;
+
+		static final String JAVA_CLASS_PATH = "java.class.path";
+		static final String PATH_SEPARATOR = "path.separator";
+		static final String DEFAULT_PATH_SEPARATOR = ":";
+
+		@Override
+		public Iterable<File> get() {
+			String classPath = System.getProperty(JAVA_CLASS_PATH, "");
+			String pathSeparator = System.getProperty(PATH_SEPARATOR, DEFAULT_PATH_SEPARATOR);
+
+			return Arrays.stream(classPath.split(pathSeparator))
+				.filter(JarsOnClassPath::notNullOrEmpty)
+				.map(File::new)
+				.filter(File::isFile)
+				.collect(Collectors.toList());
+		}
+
+		private static boolean notNullOrEmpty(String string) {
+			return !Strings.isNullOrEmpty(string);
```

Comment:
seems unnecessary to rely on an external function

Reply:
OK, changed.